processor/snapshotprocessor/factory.go

```diff
@@ -0,0 +1,127 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshotprocessor
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/processor"
+	"go.opentelemetry.io/collector/processor/processorhelper"
+)
+
+var componentType = component.MustNewType("snapshotprocessor")
+
+const (
+	stability = component.StabilityLevelAlpha
+)
+
+var consumerCapabilities = consumer.Capabilities{MutatesData: false}
+
+// NewFactory creates a new ProcessorFactory with default configuration
+func NewFactory() processor.Factory {
+	return processor.NewFactory(
+		componentType,
+		createDefaultConfig,
+		processor.WithTraces(createTracesProcessor, stability),
+		processor.WithLogs(createLogsProcessor, stability),
+		processor.WithMetrics(createMetricsProcessor, stability),
+	)
+}
+
+func createDefaultConfig() component.Config {
+	return &Config{
+		Enabled: true,
+		OpAMP:   defaultOpAMPExtensionID,
+	}
+}
+
+func createTracesProcessor(
+	ctx context.Context,
+	set processor.CreateSettings,
+	cfg component.Config,
+	nextConsumer consumer.Traces,
+) (processor.Traces, error) {
+	oCfg := cfg.(*Config)
+	sp := createOrGetProcessor(set, oCfg)
+
+	return processorhelper.NewTracesProcessor(
+		ctx,
+		set,
+		cfg,
+		nextConsumer,
+		sp.processTraces,
+		processorhelper.WithCapabilities(consumerCapabilities),
+		processorhelper.WithStart(sp.start),
+	)
+}
+
+func createLogsProcessor(
+	ctx context.Context,
+	set processor.CreateSettings,
+	cfg component.Config,
+	nextConsumer consumer.Logs,
+) (processor.Logs, error) {
+	oCfg := cfg.(*Config)
+	sp := createOrGetProcessor(set, oCfg)
+
+	return processorhelper.NewLogsProcessor(
+		ctx,
+		set,
+		cfg,
+		nextConsumer,
+		sp.processLogs,
+		processorhelper.WithCapabilities(consumerCapabilities),
+		processorhelper.WithStart(sp.start),
+	)
+}
+
+func createMetricsProcessor(
+	ctx context.Context,
+	set processor.CreateSettings,
+	cfg component.Config,
+	nextConsumer consumer.Metrics,
+) (processor.Metrics, error) {
+	oCfg := cfg.(*Config)
+	sp := createOrGetProcessor(set, oCfg)
+
+	return processorhelper.NewMetricsProcessor(
+		ctx,
+		set,
+		cfg,
+		nextConsumer,
+		sp.processMetrics,
+		processorhelper.WithCapabilities(consumerCapabilities),
+		processorhelper.WithStart(sp.start),
+	)
+}
+
+func createOrGetProcessor(set processor.CreateSettings, cfg *Config) *snapshotProcessor {
+	var sp *snapshotProcessor
+	if p, ok := processors[set.ID]; ok {
+		sp = p
+	} else {
+		sp = newSnapshotProcessor(set.Logger, cfg, set.ID)
+		processors[set.ID] = sp
+	}
+
+	return sp
+}
+
+// processors is a map of component.ID to an instance of snapshot processor.
+// It is used so that only one instance of a particular snapshot processor exists, even if it's included
+// across multiple pipelines/signal types.
+var processors = map[component.ID]*snapshotProcessor{}
```

Comment:
It's possible that components would be instantiated concurrently so I would recommend making this safe for concurrent access.